pulsar-client-cpp/lib/ReaderImpl.cc

```diff
@@ -0,0 +1,105 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "ClientImpl.h"
+#include "ReaderImpl.h"
+
+namespace pulsar {
+
+static ResultCallback emptyCallback;
+
+ReaderImpl::ReaderImpl(const ClientImplPtr client, const std::string& topic,
+                       const ReaderConfiguration& conf, const ExecutorServicePtr listenerExecutor,
+                       ReaderCallback readerCreatedCallback) :
+                topic_(topic),
+                    client_(client),
+                    readerConf_(conf),
+                    readerCreatedCallback_(readerCreatedCallback) {
+}
+
+void ReaderImpl::start(const BatchMessageId& startMessageId) {
+    ConsumerConfiguration consumerConf;
+    consumerConf.setConsumerType(ConsumerExclusive);
+    consumerConf.setReceiverQueueSize(readerConf_.getReceiverQueueSize());
+
+    if (readerConf_.getReaderName().length() > 0) {
+        consumerConf.setConsumerName(readerConf_.getReaderName());
+    }
+
+    if (readerConf_.hasReaderListener()) {
```

Comment:
A bit confused since the ReaderAPI has no method for acknowledging the message shouldn't we be calling acknowledgeIfNecessaryourselves before sending the message to ReaderListener.

Reply:
Correct, I missed that